frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/transaction_handler/editor_transaction_service.dart

```diff
@@ -0,0 +1,274 @@
+    if (node.type == type || additionalTypes.contains(node.type)) {
+      matchingNodes.add(node);
+    }
+
+    for (final child in node.children) {
+      matchingNodes.addAll(collectMatchingNodes(child, type, additionalTypes));
+    }
+
+    return matchingNodes;
+  }
+
+  void onEditorTransaction((TransactionTime, Transaction) event) {
+    if (event.$1 == TransactionTime.before) {
+      return;
+    }
+
+    final Map<String, dynamic> added = {
+      for (final handler in _transactionHandlers)
+        handler.type: handler.isParagraphSubType ? <MentionBlockData>[] : [],
+    };
+    final Map<String, dynamic> removed = {
+      for (final handler in _transactionHandlers)
+        handler.type: handler.isParagraphSubType ? <MentionBlockData>[] : [],
+    };
+
+    for (final op in event.$2.operations) {
+      if (op is InsertOperation) {
+        for (final n in op.nodes) {
+          for (final handler in _transactionHandlers) {
+            if (handler.isParagraphSubType) {
+              added[handler.type]!
+                  .addAll(extractMentionsForType(n, handler.type));
+            } else {
+              added[handler.type]!
+                  .addAll(collectMatchingNodes(n, handler.type));
+            }
+          }
+        }
+      } else if (op is DeleteOperation) {
+        for (final n in op.nodes) {
+          for (final handler in _transactionHandlers) {
+            if (handler.isParagraphSubType) {
+              removed[handler.type]!.addAll(
+                extractMentionsForType(
+                  n,
+                  handler.type,
+                  false,
+                ),
+              );
+            } else {
+              removed[handler.type]!
+                  .addAll(collectMatchingNodes(n, handler.type));
+            }
+          }
+        }
+      } else if (op is UpdateOperation) {
+        final node = widget.editorState.getNodeAtPath(op.path);
+        if (node == null) {
+          continue;
+        }
+
+        if (op.attributes['delta'] is! List ||
+            op.oldAttributes['delta'] is! List) {
+          continue;
+        }
+
+        final deltaBefore = Delta.fromJson(op.oldAttributes['delta']);
+        final deltaAfter = Delta.fromJson(op.attributes['delta']);
+
+        final (add, del) = diffDeltas(deltaBefore, deltaAfter);
+
+        for (final handler in _transactionHandlers) {
+          if (!handler.isParagraphSubType) {
+            continue;
+          }
+
+          if (add.isNotEmpty) {
+            added[handler.type]!.addAll(
+              add.where((ti) => ti.attributes?[handler.type] != null).map((ti) {
+                final index = deltaAfter.toList().indexOf(ti);
+                return (
+                  node,
+                  ti.attributes![handler.type] as Map<String, dynamic>,
+                  index,
+                );
+              }).toList(),
+            );
+          }
+
+          if (del.isNotEmpty) {
```

Comment:
Add necessary comments, and consider splitting the 'where' and 'map' into different lines. It's not easy to understand now.